Tutorial: How to run A/B tests in Android (#7194)
PostHog/posthog.com: contents/tutorials/android-ab-tests.md

---
title: "How to run A/B tests in Android with PostHog"
date: 2023-11-21
author: ["lior-neu-ner"]
tags: ['experimentation']
---

[A/B tests](/ab-testing) enables you to compare the impact of your changes on key metrics. 

PostHog makes [A/B testing on Android](/docs/experiments/installation?tab=android) simple. To show you how, this tutorial will guide you on how to run an A/B test with PostHog on an app built with Kotlin and Jetpack Compose. We will create a simple A/B test to see how the background color of a screen affects the click-through rate of a button.

## Creating a new Android app

> Our app will have two screens: The first screen will have a button which will take you to a second screen. The second screen will either have a `red` or `green` background color depending on if the user is in the `control` or `test` group. The second screen will also have a button which captures an event when it's pressed. We'll use this event as our goal metric for our test.

Reviewer: ```suggestion
Our app will have two screens: The first screen will have a button which will take you to a second screen. The second screen will either have a `red` or `green` background color depending on if the user is in the `control` or `test` group of the A/B test. The second screen will also have a button which captures an event when pressed. We'll use this event as our goal metric for our test.
```

Reviewer: Maybe a two bullet list works better here? As this is a long paragraph.

> The first step is to create a new app. Open Android studio and create a new project. Select `Empty Activity`, name your project `My-AB-Test` and use the defaults for everything else.

Reviewer: ```suggestion
The first step is to create a new app. Open [Android Studio](https://developer.android.com/studio) and create a new project. Select `Empty Activity`, name your project `My-AB-Test`, and use the defaults for everything else.
```

import androidx.navigation.compose.composable
import androidx.navigation.compose.rememberNavController
import com.example.my_ab_test.ui.theme.MyABTestTheme

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MyABTestTheme {
                Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
                    MyApp()
                }
            }
        }
    }
}

@Composable
fun MyApp() {
    val navController = rememberNavController()
    NavHost(navController = navController, startDestination = "main_screen") {
        composable("main_screen") {
            Greeting("Android", navController)
        }
        composable("second_screen") {
            SecondScreen()
        }
    }
}

@Composable
fun Greeting(name: String, navController: NavController, modifier: Modifier = Modifier) {
    Column(modifier = modifier.fillMaxSize()) {
        Text(text = "Hello $name!")
        Button(onClick = { navController.navigate("second_screen") }) {
            Text("Go to next screen")
        }
    }
}

@Composable

> fun SecondScreen() {
>     val isTestVariant = remember { mutableStateOf(false) } // We'll set this value later in the tutorial
> 
>     val backgroundColor = if (isTestVariant.value) Color.Green else Color.Red
> 
>     Surface(modifier = Modifier.fillMaxSize(), color = backgroundColor) {
> 
>         Column(modifier = Modifier.fillMaxSize(), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.Center) {

Reviewer: ```suggestion
fun SecondScreen() {
    val isTestVariant = remember { mutableStateOf(false) } // We'll set this value later in the tutorial
    val backgroundColor = if (isTestVariant.value) Color.Green else Color.Red

    Surface(modifier = Modifier.fillMaxSize(), color = backgroundColor) {
        Column(modifier = Modifier.fillMaxSize(), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.Center) {
```

Reviewer: tightening this for consistency with the rest of the code

            Button(onClick = { /* Do nothing for now */ }) {
                Text("Click Me!")
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
fun DefaultPreview() {
    MyABTestTheme {
        MyApp()
    }
}
```

> Make sure to add `implementation("androidx.navigation:navigation-compose:2.4.0")` to your dependencies in to `build.gradle (Module: app)`.

Reviewer: ```suggestion
Make sure to add `implementation("androidx.navigation:navigation-compose:2.4.0")` to your `dependencies` in to `app/build.gradle.kts` and sync your project with the Gradle files.
```

Reviewer: Weirdly when I redid the project it showed like this 😕

Reviewer: Add there is a "project view" and an "android view" need to be clear with them.

Our basic set up is now complete. Build and run your app to test that it's working.

![Basic setup of the Android app](../images/tutorials/android-ab-tests/android-basic-setup.mp4)

## Adding PostHog to your Android app

> First, add the [PostHog Android SDK](/docs/libraries/android) as a dependency in your `build.gradle` file. You can find the latest version on our [GitHub](https://github.com/PostHog/posthog-android/blob/main/CHANGELOG.md). For this tutorial, we use version `2.0.3`.

Reviewer 2: ```suggestion
First, add the [PostHog Android SDK](/docs/libraries/android) as a dependency in your `app/build.gradle.kts` file. You can find the latest version on our [GitHub](https://github.com/PostHog/posthog-android/blob/main/CHANGELOG.md). For this tutorial, we use version `2.0.3`.
```

Since there are multiple build files and this is the build file form the app itself.

Reviewer: +1

> ```gradle

Reviewer 2: ```suggestion
```kotlin
```

Reviewer 2: Since its a `build.gradle.kts` and not the `groovy` script.

dependencies {

>     implementation("com.posthog.android:posthog:2.0.3")

Reviewer: ```suggestion
    implementation("com.posthog.android:posthog:2.0.3")
    //... other dependencies
```

}
```

Sync your project with your Gradle file changes.

> Next, we create a custom application class where we can configure our PostHog instance. In a new file `MyABTestApplication.kt`, add the following code:

Reviewer: ```suggestion
Next, we create a Kotlin class where we can configure our PostHog instance. In the `src/main/java/com.example.my_ab_test` folder, add a new file `MyABTestApplication.kt`, add the following code:
```

Reviewer: It is unclear where I'm supposed to put this

Author: > It is unclear where I'm supposed to put this

I think for Android developer this is a common class to make, so I think it'll be clear for them

```kotlin
// in MyABTestApplication.kt
package com.example.my_ab_test

import android.app.Application
import com.posthog.android.PostHog

class MyABTestApplication : Application() {
    companion object {
        private const val POSTHOG_API_KEY = "<ph_project_api_key>"
        private const val POSTHOG_HOST = "<ph_instance_address>"
    }

    override fun onCreate() {
        super.onCreate()

        // Initialize PostHog
        val posthog = PostHog.Builder(this, POSTHOG_API_KEY, POSTHOG_HOST)
            .captureApplicationLifecycleEvents()
            .recordScreenViews()
            .build()

        // Set the initialized instance as a globally accessible instance
        PostHog.setSingletonInstance(posthog)
    }
}
```

> To get your PostHog API key and host, [sign up to PostHog](https://app.posthog.com/signup). Then, you can find your API key and host in your [project settings](https://app.posthog.com/settings/project). Replace the strings `<ph_project_api_key>` and `<ph_instance_address>` with your API key and host.

Reviewer: ph_project_api_key and ph_instance_address are automatically replaced in code so you don't need to mention it, and it might be confusing.
```suggestion
To get your PostHog API key and host, [sign up to PostHog](https://app.posthog.com/signup). Then, you can find your API key and host in your [project settings](https://app.posthog.com/settings/project).
```

> We now need to register our custom application class. Go to `AndroidManifest.xml` and add `android:name=".MyABTestApplication"` within the `<application` tag. We also need to give our app internet permission by adding `<uses-permission android:name="android.permission.INTERNET" />`. 

Reviewer 2: Did you run into internet permission issues?
Usually this is not needed since the Android SDK already has the internet permission and they are merged during the build process.
https://github.com/PostHog/posthog-android/blob/b8866efac427a08646c0f17c57c2b6c2df1b17ee/posthog/src/main/AndroidManifest.xml#L31

Author: Yeah I did. App was crashing on launch because of this

Together both of these changes look like this:

```XML
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools">

    <uses-permission android:name="android.permission.INTERNET" />

    <application
        android:name=".MyABTestApplication"
        ...
    </application>
</manifest>
```

Finally, we [capture a custom event](/docs/libraries/android#capturing-events) when the button on `SecondScreen` is clicked. We'll use this event as our goal metric in our A/B test.

Update the `SecondScreen` function in `MainActivity.kt` to the following:

```kotlin
// in MainActivity.kt

import com.posthog.android.PostHog
import androidx.compose.ui.platform.LocalContext
// ... [Rest of your imports and MainActivity code]

@Composable
fun SecondScreen() {
    val isTestVariant = remember { mutableStateOf(false) } // We'll set this value later in the tutorial
    val backgroundColor = if (isTestVariant.value) Color.Green else Color.Red

>     val context = LocalContext.current

Reviewer: ```suggestion
    val context = LocalContext.current // new
```

    Surface(modifier = Modifier.fillMaxSize(), color = backgroundColor) {

> 

Reviewer: ```suggestion
```

        Column(modifier = Modifier.fillMaxSize(), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.Center) {
            Button(onClick = {
                PostHog.with(context)
                    .capture("feature_button_clicked");
            }) {
                Text("Click Me!")
            }
        }
    }
}
```

> To check your setup, build and run your app. Click your button a few times. You should start seeing events in the [PostHog event explorer](https://app.posthog.com/events).

Reviewer: ```suggestion
To check your setup, build and run your app. Click your button a few times and you should start seeing events in the [PostHog event explorer](https://app.posthog.com/events).
```

![Android events captured](../images/tutorials/android-ab-tests/android-events.png)

## Create an A/B test in PostHog

The next step is to set up an A/B test (we call them experiments in PostHog).

Go to the [Experiments tab](https://app.posthog.com/experiments) in PostHog and click "New experiment". Add the following details to your experiment:

1. Name it "Android background color experiment".
2. Set "Feature flag key" to `android-background-color-experiment`.
3. Under the experiment goal, select the `feature_button_clicked` we created in the previous step.
4. Use the default values for all other fields.

Click "Save as draft" and then click "Launch".

![Experiment setup in PostHog](../images/tutorials/android-ab-tests/experiment-setup.png)

## Implement the A/B test code

> The final step is to add the experiment code. We'll add code that does the following:

Reviewer: ```suggestion
The final step is to add the experiment code that does the following:
```
